### PR TITLE
Optimizes memory usage of AliasCache

### PR DESCRIPTION
**Pull request by balazsracz**
Changes the storage datastructure in AliasCache to use less memory.

The new structure uses 16-bit indexes instead of pointers, and a
sorted array of 16-bit indexes instead of red-black trees for the
lookup methods.
The LRU double-linked list is kept.

With this change the memory cost goes from 88 bytes per entry to
16 bytes per entry. The compiled code size is also smaller.

### CODE REVIEW

File: src/openlcb/AliasCache.cxx [hunk 1]
```diff
@@ -185,9 +185,9 @@ bool AliasCache::retrieve(unsigned entry, NodeID* node, NodeAlias* alias)
 {
     HASSERT(entry < size());
     Metadata* md = pool + entry;
-    if (!md->alias) return false;
-    if (node) *node = md->id;
-    if (alias) *alias = md->alias;
+    if (!md->alias_) return false;
```
Comment (bakerstu) [inline src/openlcb/AliasCache.cxx:188]: Should we take the opportunity of touching this to fix the style: All if statements must have braces.

File: src/openlcb/AliasCache.cxxtest [hunk 2]
```diff
@@ -486,66 +488,74 @@ int AliasCache::check_consistency() {
...
-        if (free_entries.count(kv.second)) {
+        if (free_entries.count(kv.deref(this)))
+        {
             return 20;
         }
     }
     if (aliasMap.size() == 0) {
-        if (oldest != nullptr) return 7;
-        if (newest != nullptr) return 8;
+        if (!oldest.empty()) return 7;
```
Comment (bakerstu) [inline src/openlcb/AliasCache.cxxtest:503]: same.

File: src/openlcb/AliasCache.cxxtest [hunk 3]
```diff
@@ -486,66 +488,74 @@ int AliasCache::check_consistency() {
...
+        {
+            auto next = prev.deref(this)->newer_;
             ++count;
-            if (free_entries.count(next)) {
+            if (free_entries.count(next.deref(this)))
+            {
                 return 21;
             }
-            if (next->older != prev) return 14;
+            if (next.deref(this)->older_.idx_ != prev.idx_) return 14;
```
Comment (bakerstu) [inline src/openlcb/AliasCache.cxxtest:531]: same.

File: src/openlcb/AliasCache.cxxtest [hunk 4]
```diff
@@ -486,66 +488,74 @@ int AliasCache::check_consistency() {
...
     if (aliasMap.size() == 0) return 0;
     // Check linking.
     {
-        Metadata* prev = oldest;
+        PoolIdx prev = oldest;
         unsigned count = 1;
-        if (prev->older) return 13;
-        while (prev->newer) {
-            auto* next = prev->newer;
+        if (!prev.deref(this)->older_.empty()) return 13;
```
Comment (bakerstu) [inline src/openlcb/AliasCache.cxxtest:522]: same.

File: src/openlcb/AliasCache.cxxtest [hunk 5]
```diff
@@ -486,66 +488,74 @@ int AliasCache::check_consistency() {
...
         if (count != aliasMap.size()) return 27;
     }
     {
-        Metadata* next = newest;
-        if (next->newer) return 15;
-        while (next->older) {
-            auto* prev = next->older;
-            if (free_entries.count(prev)) {
+        PoolIdx next = newest;
+        if (!next.deref(this)->newer_.empty()) return 15;
```
Comment (bakerstu) [inline src/openlcb/AliasCache.cxxtest:539]: same.

File: src/openlcb/AliasCache.cxxtest [hunk 6]
```diff
@@ -486,66 +488,74 @@ int AliasCache::check_consistency() {
...
+        if (next.idx_ != oldest.idx_) return 17;
     }
     for (unsigned i = 0; i < entries; ++i) {
         if (free_entries.count(pool+i)) continue;
         auto* e = pool+i;
-        if (idMap.find(e->id) == idMap.end()) return 23;
-        if (idMap[e->id] != e) return 24;
-        if (aliasMap.find(e->alias) == aliasMap.end()) return 25;
-        if (aliasMap[e->alias] != e) return 26;
+        if (idMap.find(e->get_node_id()) == idMap.end()) return 23;
```
Comment (bakerstu) [inline src/openlcb/AliasCache.cxxtest:555]: same.

**Review by bakerstu — APPROVED** [review]

File: src/openlcb/AliasCache.cxx (hunk 1 above)
Comment (balazsracz) [reply src/openlcb/AliasCache.cxx:188]: All of these style issues are fixed in https://github.com/bakerstu/openmrn/pull/433.

**Pull request merged**